debian/docs/sb_upload.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.10.
.TH SB_UPLOAD "1" "June 2014" "sb_upload " "User Commands"
.SH NAME
sb_upload - Read a message or a mailbox file, upload it to a web server and write it
.SH DESCRIPTION
Read a message or a mailbox file on standard input, upload it to a
web server and write it to standard output.
.PP
By default, this sends the message to the SpamBayes sb_server web
interface, which will save the message in the 'unknown' cache, ready
for you to classify it.  It does not do any training, just saves it
ready for you to classify (unless you use the \fB\-t\fR switch).
.PP
usage:  /usr/bin/sb_upload [\-h] [\-n] [\-s server] [\-p port] [\-r N]
.IP
[\-o section:option:value]
[\-t (ham|spam)]
.HP
\fB\-h\fR, \fB\-\-help\fR    \- print help and exit
.HP
\fB\-n\fR, \fB\-\-null\fR    \- suppress writing to standard output (default False)
.HP
\fB\-s\fR, \fB\-\-server=\fR \- provide alternate web server (default localhost)
.HP
\fB\-p\fR, \fB\-\-port=\fR   \- provide alternate server port (default 8880)
.HP
\fB\-r\fR, \fB\-\-prob=\fR   \- feed the message to the trainer w/ prob N [0.0...1.0]
.HP
\fB\-t\fR, \fB\-\-train=\fR  \- train the message (pass either 'ham' or 'spam')
.HP
\fB\-o\fR, \fB\-\-option=\fR \- set [section, option] in the options database to value
.SH "SEE ALSO"
The full documentation for
.B sb_upload
is maintained as a Texinfo manual.  If the
.B info
and
.B sb_upload
programs are properly installed at your site, the command
.IP
.B info sb_upload
.PP
should give you access to the complete manual.


CHANGELOG.txt
[Note that all dates are in English, not American format - i.e. day/month/year]

Release 1.0.4
=============
Tony Meyer        09/03/2005  Use a blocking socket when connecting with sb_server, so that with Python 2.4 we don't end up with never-ending reports of errors.
Tony Meyer        07/02/2005  sb_server: When non-verbose, only report unique connection errors once per hour.

Release 1.0.3
=============
Tony Meyer        28/01/2005  Distribute mscvr71.dll, as Microsoft recommend.
Tony Meyer        28/01/2005  The 'fix' for escaping subject lines was incorrect; do it correctly.

Release 1.0.2
=============
Tony Meyer        14/01/2005  Try and work around the reasonably common problems with imaplib and OS X.
Tony Meyer        04/01/2005  If we are reloading the options, optionsPathname will already be set, so if we are  a windows user using the default pathname, the object won't be reloaded.  Fix so a reload is always done.
Tony Meyer        02/01/2005  UserInterface.py: cgi.escape the configuration filename
Tony Meyer        22/12/2004  Make Dibbler.py compatible with Python 2.2.1.
Tony Meyer        17/12/2004  message.py: Change the deprecated function to work in Python 2.4.
Tony Meyer        06/12/2004  Fix [ 1078923 ] Unicode support incomplete
Tony Meyer        06/12/2004  Fix the regex that the auth digest used, and handle the auth digest responses tha IE 6.0 and Firefox 1.0 give.
Tony Meyer        29/11/2004  Subject lines are not cgi.escape()d in the web interface, which might cause errors - fixed.

Release 1.0.1
=============
Tony Meyer        15/11/2004  Fix a bug in sb_dbexpimp.py where merging into an existing dbm file might lose training data.
Tony Meyer        15/11/2004  sb_dbexpimp.py: Fail if the csv file doesn't exist that we are trying to import from rather than  keeping going, which made no sense.
Tony Meyer        11/11/2004  The installer wasn't offered to install a startup items shortcut, so fix that.  This is a non-ideal patch, but appears to be the only way Inno will work.
Tony Meyer        03/11/2004  Fix [ 1022848 ] sb_dbexpimp.py crashes while importing into pickle file
Tony Meyer        03/11/2004  Fix [ 831864 ] sb_mboxtrain.py: flock vs. lockf
Tony Meyer        03/11/2004  Fix [ 922063 ] Intermittent sb_filter.py failure with URL pickle
Tony Meyer        29/09/2004  Fix [ 1036601 ] typo on advanced config web page
Tony Meyer        28/10/2004  Add [ 715248 ] Pickle classifier should save to a temp file first
Tony Meyer        21/10/2004  Fix [ 1051081 ] uncaught socket timeoutexception slurping URLs
Tony Meyer        18/10/2004  TestDriver: If show_histograms was False, then the global ham/spam histogram never had the stats computed, but this gets used later, so the script would die with an AtrributeError. Fix that.
Tony Meyer        13/10/2004  Fix mySQL storage option for the case where the server does not support rollbacks.
Sjoerd Mullender  02/10/2004  imapfilter: Quote the search string that tries to find the message again that was just saved.
Tony Meyer        30/09/2004  Fix [ 903905 ] IMAP Configuration Error
Tony Meyer        15/09/2004  sb_upload: Clarify docstring so that it's more clear what this script does. The -n / --null command line option didn't actually do anything; change it so that it does.
Tony Meyer        23/07/2004  For proxy handler for version checking, the proxy port needs to be an integer, not a string.
Tony Meyer        19/07/2004  Fix [ 990700 ] Changes to asyncore in Python 2.4 break ServerLineReader
Kenny Pitt        17/07/2004  Fix [941639] and [986353].  Use a non-standard extension for our py2exe created zip to get around Windows extensions that automatically expand zip files.
Tony Meyer        14/07/2004  Fix [ 790757 ] signal handler created with wrong # of args
Tony Meyer        14/07/2004  Fix [ 944109 ] notate_to/subject option valid values should be dynamic
Tony Meyer        14/07/2004  Fix [ 959937 ] "Invalid server" message not always correct
Skip Montanaro    10/07/2004  tte.py: 2.3 compatibility: add reversed() function
Tony Meyer        09/07/2004  Update test_storage.py test to reflect (current) correct way to call open_storage. Fixes part of [ 981970 ] tests failing.
Tony Meyer        09/07/2004  Using -u with sb_server had been broken.  Fix this.

1.0 Final
=========
(no changes from 1.0rc2)

Release Candidate 2
===================
Tony Meyer        10/06/2004  Install license with binaries.
Mark Hammond      01/06/2004  Fix Tony's fix for [ 943397 ] Cannot choose more than one Inbox folder to filter
Tony Meyer        02/06/2004  Fix [ 961713 ] server error when trying to change Statistics Options
Skip Montanaro    31/05/2004  Compatcsv: Abandon regexes as bad way to parse csv file - add simple unit test as well
Skip Montanaro    31/05/2004  sb_dbexpimp.py: handle case of importing the old object craft csv module.
Skip Montanaro    22/05/2004  sb_dbexpimp.py: Add a fake UnicodeDecodeError for Python 2.2 compatibility.
Skip Montanaro    22/05/2004  Compatcsv: correct a list chomping bug (forgot a slicing ":") and use .readline() instead of .next() for Python 2.2 compatibility
Tony Meyer        18/05/2004  Fix [ 955442 ] default_bayes_customize.ini not handled right
Tony Meyer        14/05/2004  message.py: If we trigger the TypeError problem, try to do our best to put the message together anyway.
Tony Meyer        14/05/2004  Handle people using a storage type that isn't pickle/dbm (like one of the sql ones).
Tony Meyer        13/05/2004  If a user tried to send a bug report before entering in any servers to proxy, this would generate an IndexError.  Fix that.
Tony Meyer        12/05/2004  Fix [ 943397 ] Cannot choose more than one Inbox folder to filter

Release Candidate 1
===================
Skip Montanaro    05/05/2004  sb_filter, sb_bnfilter use new mboxutils as_string().

scripts/sb_upload.py
#!/usr/bin/env python

"""
Read a message or a mailbox file on standard input, upload it to a
web server and write it to standard output.

By default, this sends the message to the SpamBayes sb_server web
interface, which will save the message in the 'unknown' cache, ready
for you to classify it.  It does not do any training, just saves it
ready for you to classify.

usage:  %(progname)s [-h] [-n] [-s server] [-p port] [-r N]
                     [-o section:option:value]

Options:
    -h, --help    - print help and exit
    -n, --null    - suppress writing to standard output (default %(null)s)
    -s, --server= - provide alternate web server (default %(server)s)
    -p, --port=   - provide alternate server port (default %(port)s)
    -r, --prob=   - feed the message to the trainer w/ prob N [0.0...1.0]
    -o, --option= - set [section, option] in the options database to value
"""

import sys
import httplib
import mimetypes
import getopt
import random
from spambayes.Options import options

progname = sys.argv[0]

__author__ = "Skip Montanaro <skip@pobox.com>"
__credits__ = "Spambayes gang, Wade Leftwich"

try:
    True, False
except NameError:
    # Maintain compatibility with Python 2.2
    True, False = 1, 0

# appropriated verbatim from a recipe by Wade Leftwich in the Python
# Cookbook: http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/146306

def post_multipart(host, selector, fields, files):
    """
    Post fields and files to an http host as multipart/form-data.  fields is
    a sequence of (name, value) elements for regular form fields.  files is
    a sequence of (name, filename, value) elements for data to be uploaded
    as files.  Return the server's response page.
    """
    content_type, body = encode_multipart_formdata(fields, files)
    h = httplib.HTTP(host)
    h.putrequest('POST', selector)
    h.putheader('content-type', content_type)
    h.putheader('content-length', str(len(body)))
    h.endheaders()
    h.send(body)
    errcode, errmsg, headers = h.getreply()
    return h.file.read()

def encode_multipart_formdata(fields, files):
    """
    fields is a sequence of (name, value) elements for regular form fields.
    files is a sequence of (name, filename, value) elements for data to be
    uploaded as files.  Return (content_type, body) ready for httplib.HTTP
    instance
    """
    BOUNDARY = '----------ThIs_Is_tHe_bouNdaRY_$'
    CRLF = '\r\n'
    L = []
    for (key, value) in fields:
        L.append('--' + BOUNDARY)
        L.append('Content-Disposition: form-data; name="%s"' % key)
        L.append('')
        L.append(value)
    for (key, filename, value) in files:
        L.append('--' + BOUNDARY)
        L.append('Content-Disposition: form-data; name="%s"; filename="%s"' % (key, filename))
        L.append('Content-Type: %s' % get_content_type(filename))
        L.append('')
        L.append(value)
    L.append('--' + BOUNDARY + '--')
    L.append('')
    body = CRLF.join(L)
    content_type = 'multipart/form-data; boundary=%s' % BOUNDARY
    return content_type, body

def get_content_type(filename):
    return mimetypes.guess_type(filename)[0] or 'application/octet-stream'

def usage(*args):
    defaults = {}
    for d in args:
        defaults.update(d)
    print __doc__ % defaults

def main(argv):
    null = False
    server = "localhost"
    port = options["html_ui", "port"]
    prob = 1.0

    try:
        opts, args = getopt.getopt(argv, "hns:p:r:o:",
                                   ["help", "null", "server=", "port=",
                                    "prob=", "option="])
    except getopt.error:
        usage(globals(), locals())
        sys.exit(1)

    for opt, arg in opts:
        if opt in ("-h", "--help"):
            usage(globals(), locals())
            sys.exit(0)
        elif opt in ("-n", "--null"):
            null = True
        elif opt in ("-s", "--server"):
            server = arg
        elif opt in ("-p", "--port"):
            port = int(arg)
        elif opt in ("-r", "--prob"):
            n = float(arg)
            if n < 0.0 or n > 1.0:
                usage(globals(), locals())
                sys.exit(1)
            prob = n
        elif opt in ('-o', '--option'):
            options.set_from_cmdline(arg, sys.stderr)

    if args:
        usage(globals(), locals())
        sys.exit(1)

    data = sys.stdin.read()
    if not null:
        sys.stdout.write(data)
    if random.random() < prob:
        try:
            post_multipart("%s:%d"%(server,port), "/upload", [],
                               [('file', 'message.dat', data)])
        except:
            # not an error if the server isn't responding
            pass

if __name__ == "__main__":
    main(sys.argv[1:])
